src/libopensc/pkcs15.c

```diff
@@ -2432,6 +2475,46 @@ sc_pkcs15_read_file(struct sc_pkcs15_card *p15card, const struct sc_path *in_pat
 			}
 			len = head-data;
 		}
+		else if (file->ef_structure == SC_FILE_EF_LINEAR_VARIABLE) {
+
+			u8 offset_buffer[] = {0x54, 0x02, 0x00, 0x00};
+			struct sc_apdu apdu;
+			uint16_t offset_u16 = 0;
+			u8 *data_do;
+			size_t data_do_len;
+
+			do {
+
+				offset_buffer[2] = (u8) (offset_u16 >> 8);
+				offset_buffer[3] = (u8) (offset_u16 & 0x00ff);
+
+				sc_format_apdu(p15card->card, &apdu, SC_APDU_CASE_4_SHORT, 0xB3, in_path->index, 4);
+				
+				apdu.data = offset_buffer;
+				apdu.datalen = 4;
+				apdu.lc = 4;
+				apdu.le = 0;
+				
+				apdu.resp = data + offset_u16;
+				apdu.resplen = len - offset_u16;
+	
+				r = sc_transmit_apdu(p15card->card, &apdu);
+				if(r < 0 || apdu.resplen == 0) break;
+
+				data_do = NULL;
+				r = sc_decode_do53(ctx, &data_do, &data_do_len,	apdu.resp, apdu.resplen);
+				if (r < 0) goto fail_unlock;
+
+				if(data_do) {
+					memcpy(data + offset_u16, data_do, data_do_len);
```

Comment:
I assume you can do this because you added +5 to the data allocated above.

I would consider reading the data into buffer of `SC_MAX_APDU_RESP_SIZE` size, do the decoding in there and only after that copy the result to the `data` buffer.

Reply:
What I think is more important, @wcosnc correctly identified that `sc_read_record` (or its counterpart `read_record` in the card driver call backs) doesn't allow specifying an offset. Doing this here, is bad since we want to have the flexibility of implementing this depending on the card driver in use.

Currently, only iso7816.c has an implementation of read_record so changing this would have very minimal impact on the other cards. I suggest you do the following:
1. add an additional parameter to both sc_read_record and read_record, named idx (just like in sc_read_binary)
2. Use idx = 0 where sc_read_record is currently called without this parameter
3. change read_record in iso7816.c so that if and only if  `idx != 0` your implementation is used with INS=B3. if idx is not specified (i.e. 0), then it should still use the current version with INS=B2
4. use the new sc_read_record to specify the offset in sc_pkcs15_read_file